src/lcm/discrete_problem.py

```diff
@@ -0,0 +1,280 @@
+
+"""
+
+from collections.abc import Callable
+from functools import partial
+from typing import Any, Literal
+
+import jax
+import jax.numpy as jnp
+import pandas as pd
+from jax import Array
+
+from lcm.typing import SegmentInfo
+
+
+def get_solve_discrete_problem(
+    *,
+    shock_type: Literal["extreme_value"] | None,
+    variable_info: pd.DataFrame,
+    is_last_period: bool,
+    choice_segments: SegmentInfo | None,
+    params: dict[str, Any],
+) -> Callable[[Array], Array]:
+    """Get function that computes the expected max. of conditional continuation values.
+
+    The maximum is taken over the discrete and sparse choice variables in each state.
+
+    Args:
+        shock_type (Literal["extreme_value"] | None): Type of choice shock. Currently
+            only None is supported. Work for "extreme_value" is in progress.
+        variable_info (pd.DataFrame): DataFrame with information about the variables.
+        is_last_period (bool): Whether the function is created for the last period.
+        choice_segments (SegmentInfo): Dictionary with the entries "segment_ids" and
+            "num_segments". segment_ids are a 1d integer array that partitions the first
+            dimension of values into choice sets over which we need to aggregate.
+            "num_segments" is the number of choice sets. The segment_ids are assumed to
+            be sorted.
+        params (dict): Dictionary with model parameters.
+
+    Returns:
+        callable: Function that calculates the expected maximum of the conditional
+            continuation values. The function depends on `cc_values` (jax.Array), the
+            conditional continuation values, and returns the aggregated values.
+
+    """
+    if is_last_period:
+        variable_info = variable_info.query("~is_auxiliary")
+
+    choice_axes = _determine_dense_discrete_choice_axes(variable_info)
+
+    if shock_type is None:
+        func = _solve_discrete_problem_no_shocks
+    elif shock_type == "extreme_value":
+        raise NotImplementedError("Extreme value shocks are not yet implemented.")
+    else:
+        raise ValueError(f"Invalid shock_type: {shock_type}.")
+
+    return partial(
+        func,
+        choice_axes=choice_axes,
+        choice_segments=choice_segments,
+        params=params,
+    )
+
+
+# ======================================================================================
+# Discrete problem with no shocks
+# ======================================================================================
+
+
+def _solve_discrete_problem_no_shocks(
+    cc_values: Array,
+    choice_axes: tuple[int, ...] | None,
+    choice_segments: SegmentInfo | None,
+    params: dict[str, Any],  # noqa: ARG001
+) -> Array:
+    """Aggregate conditional continuation values over discrete choices.
+
+    Args:
+        cc_values (jax.Array): Array with conditional continuation values. Has one
+            dimensions per state and discrete choice variable.
```

Comment:
```suggestion
            state and discrete choice variable, it has one axis.
```
Not beautiful, but a little clearer this way around.